File: stock-quantity-zero-despite-inventory/python/detect_stock_drift.py
"""Find Saleor variant and warehouse pairs where Stock.quantity does not
match reality: either live Allocation rows exceed it, a known physical
count from a WMS export is higher than it, or it is zero while open
allocations exist (saleor/saleor#5578, #4058, #543).

This script never overwrites inventory on its own. Under DRY_RUN=true
(the default) it only reports drifted pairs. When DRY_RUN=false and a
confirmed physical count is supplied per variant and warehouse, it
applies the correction one pair at a time and re-checks quantityAvailable
before moving on. Run on a schedule. Safe to run again and again.

Guide: https://www.allanninal.dev/saleor/stock-quantity-zero-despite-inventory/
"""
import os
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("detect_stock_drift")

API_URL = os.environ.get("SALEOR_API_URL", "https://store.saleor.cloud/graphql/")
TOKEN = os.environ.get("SALEOR_AUTH_TOKEN", "dummy-token")
CHANNEL = os.environ.get("SALEOR_CHANNEL", "default-channel")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

WAREHOUSES_QUERY = """
query($cursor: String) {
  warehouses(first: 50, after: $cursor) {
    pageInfo { hasNextPage endCursor }
    edges { node { id name slug } }
  }
}"""

VARIANTS_QUERY = """
query($channel: String!, $cursor: String) {
  productVariants(channel: $channel, first: 50, after: $cursor) {
    pageInfo { hasNextPage endCursor }
    edges {
      node {
        id
        sku
        quantityAvailable(countryCode: US)
        stocks { warehouse { id slug } quantity quantityAllocated }
      }
    }
  }
}"""

ORDERS_WITH_ALLOCATIONS_QUERY = """
query($cursor: String) {
  orders(first: 50, after: $cursor,
         filter: { status: [UNFULFILLED, PARTIALLY_FULFILLED] }) {
    pageInfo { hasNextPage endCursor }
    edges {
      node {
        id
        lines {
          variant { id sku }
          allocations { quantity warehouse { id } }
        }
      }
    }
  }
}"""

STOCK_BULK_UPDATE = """
mutation($variantId: ID!, $warehouseId: ID!, $quantity: Int!) {
  stockBulkUpdate(stocks: [{ variantId: $variantId, warehouseId: $warehouseId, quantity: $quantity }]) {
    results { stock { id quantity } errors { field message code } }
  }
}"""

VARIANT_AVAILABILITY_QUERY = """
query($id: ID!, $channel: String!) {
  productVariant(id: $id, channel: $channel) { id quantityAvailable(countryCode: US) }
}"""


def gql(query, variables=None):
    r = requests.post(
        API_URL,
        json={"query": query, "variables": variables or {}},
        headers={"Authorization": f"Bearer {TOKEN}", "Content-Type": "application/json"},
        timeout=30,
    )
    r.raise_for_status()
    body = r.json()
    if body.get("errors"):
        raise RuntimeError(body["errors"])
    return body["data"]


def detect_stock_drift(stock, allocations, known_physical_count=None):
    """Pure decision function: no network calls.

    stock: {quantity, quantityAllocated, variantId, warehouseId}
    allocations: [{quantity}, ...]
    known_physical_count: int | None

    Returns {isDrift, delta, reason} where reason is one of
    "allocated_exceeds_quantity", "quantity_below_known_physical_count",
    or "zero_quantity_with_open_allocations".
    """
    allocated_sum = sum(a["quantity"] for a in allocations)
    quantity = stock["quantity"]

    if quantity == 0 and allocated_sum > 0:
        return {"isDrift": True, "delta": allocated_sum, "reason": "zero_quantity_with_open_allocations"}

    if allocated_sum > quantity:
        return {"isDrift": True, "delta": allocated_sum - quantity, "reason": "allocated_exceeds_quantity"}

    if known_physical_count is not None and known_physical_count > quantity:
        return {"isDrift": True, "delta": known_physical_count - quantity, "reason": "quantity_below_known_physical_count"}

    return {"isDrift": False, "delta": 0, "reason": ""}


def all_warehouses():
    cursor = None
    rows = []
    while True:
        data = gql(WAREHOUSES_QUERY, {"cursor": cursor})["warehouses"]
        rows.extend(edge["node"] for edge in data["edges"])
        if not data["pageInfo"]["hasNextPage"]:
            return rows
        cursor = data["pageInfo"]["endCursor"]


def stock_snapshot(channel):
    cursor = None
    rows = []
    while True:
        data = gql(VARIANTS_QUERY, {"channel": channel, "cursor": cursor})["productVariants"]
        for edge in data["edges"]:
            node = edge["node"]
            for stock in node["stocks"]:
                rows.append({
                    "variantId": node["id"],
                    "sku": node["sku"],
                    "warehouseId": stock["warehouse"]["id"],
                    "quantity": stock["quantity"],
                    "quantityAllocated": stock["quantityAllocated"],
                })
        if not data["pageInfo"]["hasNextPage"]:
            return rows
        cursor = data["pageInfo"]["endCursor"]


def allocations_for(variant_id, warehouse_id):
    cursor = None
    matches = []
    while True:
        data = gql(ORDERS_WITH_ALLOCATIONS_QUERY, {"cursor": cursor})["orders"]
        for edge in data["edges"]:
            for line in edge["node"]["lines"]:
                if not line["variant"] or line["variant"]["id"] != variant_id:
                    continue
                for allocation in line["allocations"]:
                    if allocation["warehouse"]["id"] == warehouse_id:
                        matches.append({"quantity": allocation["quantity"]})
        if not data["pageInfo"]["hasNextPage"]:
            return matches
        cursor = data["pageInfo"]["endCursor"]


def apply_correction(variant_id, warehouse_id, corrected_quantity):
    result = gql(STOCK_BULK_UPDATE, {
        "variantId": variant_id,
        "warehouseId": warehouse_id,
        "quantity": corrected_quantity,
    })["stockBulkUpdate"]
    for item in result["results"]:
        if item["errors"]:
            raise RuntimeError(item["errors"])
    return result


def confirm_available(variant_id, channel):
    data = gql(VARIANT_AVAILABILITY_QUERY, {"id": variant_id, "channel": channel})["productVariant"]
    return data["quantityAvailable"]


def run(known_physical_counts=None):
    known_physical_counts = known_physical_counts or {}
    stocks = stock_snapshot(CHANNEL)
    flagged = []

    for stock in stocks:
        allocations = allocations_for(stock["variantId"], stock["warehouseId"])
        known = known_physical_counts.get((stock["variantId"], stock["warehouseId"]))
        result = detect_stock_drift(stock, allocations, known)
        if not result["isDrift"]:
            continue
        flagged.append({**stock, **result, "suspectedPhysicalCount": known})
        log.warning(
            "DRIFT sku=%s variant=%s warehouse=%s quantity=%d allocated=%d reason=%s delta=%d",
            stock["sku"], stock["variantId"], stock["warehouseId"],
            stock["quantity"], stock["quantityAllocated"], result["reason"], result["delta"],
        )

    if DRY_RUN:
        log.info("Done (dry run). %d drifted variant and warehouse pair(s) reported.", len(flagged))
        return flagged

    for row in flagged:
        known = row["suspectedPhysicalCount"]
        if known is None:
            log.info("Skipping %s at %s, no confirmed physical count supplied.", row["sku"], row["warehouseId"])
            continue
        apply_correction(row["variantId"], row["warehouseId"], known)
        available = confirm_available(row["variantId"], CHANNEL)
        log.info("Corrected %s at %s to %d. quantityAvailable now %s.", row["sku"], row["warehouseId"], known, available)

    log.info("Done. %d drifted variant and warehouse pair(s) processed.", len(flagged))
    return flagged


if __name__ == "__main__":
    run()

File: stock-quantity-zero-despite-inventory/node/detect-stock-drift.js
/**
 * Find Saleor variant and warehouse pairs where Stock.quantity does not
 * match reality: either live Allocation rows exceed it, a known physical
 * count from a WMS export is higher than it, or it is zero while open
 * allocations exist (saleor/saleor#5578, #4058, #543).
 *
 * This script never overwrites inventory on its own. Under DRY_RUN=true
 * (the default) it only reports drifted pairs. When DRY_RUN=false and a
 * confirmed physical count is supplied per variant and warehouse, it
 * applies the correction one pair at a time and re-checks quantityAvailable
 * before moving on. Run on a schedule.
 *
 * Guide: https://www.allanninal.dev/saleor/stock-quantity-zero-despite-inventory/
 */
import { pathToFileURL } from "node:url";

const API_URL = process.env.SALEOR_API_URL || "https://store.saleor.cloud/graphql/";
const TOKEN = process.env.SALEOR_AUTH_TOKEN || "dummy-token";
const CHANNEL = process.env.SALEOR_CHANNEL || "default-channel";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

/**
 * Pure decision function: no network calls.
 *
 * stock: {quantity, quantityAllocated, variantId, warehouseId}
 * allocations: [{quantity}, ...]
 * knownPhysicalCount: number | null
 *
 * Returns {isDrift, delta, reason} where reason is one of
 * "allocated_exceeds_quantity", "quantity_below_known_physical_count",
 * or "zero_quantity_with_open_allocations".
 */
export function detectStockDrift(stock, allocations, knownPhysicalCount = null) {
  const allocatedSum = allocations.reduce((sum, a) => sum + a.quantity, 0);
  const quantity = stock.quantity;

  if (quantity === 0 && allocatedSum > 0) {
    return { isDrift: true, delta: allocatedSum, reason: "zero_quantity_with_open_allocations" };
  }

  if (allocatedSum > quantity) {
    return { isDrift: true, delta: allocatedSum - quantity, reason: "allocated_exceeds_quantity" };
  }

  if (knownPhysicalCount !== null && knownPhysicalCount > quantity) {
    return { isDrift: true, delta: knownPhysicalCount - quantity, reason: "quantity_below_known_physical_count" };
  }

  return { isDrift: false, delta: 0, reason: "" };
}

async function gql(query, variables = {}) {
  const res = await fetch(API_URL, {
    method: "POST",
    headers: { Authorization: `Bearer ${TOKEN}`, "Content-Type": "application/json" },
    body: JSON.stringify({ query, variables }),
  });
  if (!res.ok) throw new Error(`Saleor ${res.status}`);
  const body = await res.json();
  if (body.errors) throw new Error(JSON.stringify(body.errors));
  return body.data;
}

const WAREHOUSES_QUERY = `
query($cursor: String) {
  warehouses(first: 50, after: $cursor) {
    pageInfo { hasNextPage endCursor }
    edges { node { id name slug } }
  }
}`;

const VARIANTS_QUERY = `
query($channel: String!, $cursor: String) {
  productVariants(channel: $channel, first: 50, after: $cursor) {
    pageInfo { hasNextPage endCursor }
    edges {
      node {
        id
        sku
        quantityAvailable(countryCode: US)
        stocks { warehouse { id slug } quantity quantityAllocated }
      }
    }
  }
}`;

const ORDERS_WITH_ALLOCATIONS_QUERY = `
query($cursor: String) {
  orders(first: 50, after: $cursor,
         filter: { status: [UNFULFILLED, PARTIALLY_FULFILLED] }) {
    pageInfo { hasNextPage endCursor }
    edges {
      node {
        id
        lines {
          variant { id sku }
          allocations { quantity warehouse { id } }
        }
      }
    }
  }
}`;

const STOCK_BULK_UPDATE = `
mutation($variantId: ID!, $warehouseId: ID!, $quantity: Int!) {
  stockBulkUpdate(stocks: [{ variantId: $variantId, warehouseId: $warehouseId, quantity: $quantity }]) {
    results { stock { id quantity } errors { field message code } }
  }
}`;

const VARIANT_AVAILABILITY_QUERY = `
query($id: ID!, $channel: String!) {
  productVariant(id: $id, channel: $channel) { id quantityAvailable(countryCode: US) }
}`;

async function allWarehouses() {
  let cursor = null;
  const rows = [];
  while (true) {
    const data = (await gql(WAREHOUSES_QUERY, { cursor })).warehouses;
    rows.push(...data.edges.map((edge) => edge.node));
    if (!data.pageInfo.hasNextPage) return rows;
    cursor = data.pageInfo.endCursor;
  }
}

async function stockSnapshot(channel) {
  let cursor = null;
  const rows = [];
  while (true) {
    const data = (await gql(VARIANTS_QUERY, { channel, cursor })).productVariants;
    for (const edge of data.edges) {
      const node = edge.node;
      for (const stock of node.stocks) {
        rows.push({
          variantId: node.id,
          sku: node.sku,
          warehouseId: stock.warehouse.id,
          quantity: stock.quantity,
          quantityAllocated: stock.quantityAllocated,
        });
      }
    }
    if (!data.pageInfo.hasNextPage) return rows;
    cursor = data.pageInfo.endCursor;
  }
}

async function allocationsFor(variantId, warehouseId) {
  let cursor = null;
  const matches = [];
  while (true) {
    const data = (await gql(ORDERS_WITH_ALLOCATIONS_QUERY, { cursor })).orders;
    for (const edge of data.edges) {
      for (const line of edge.node.lines) {
        if (!line.variant || line.variant.id !== variantId) continue;
        for (const allocation of line.allocations) {
          if (allocation.warehouse.id === warehouseId) matches.push({ quantity: allocation.quantity });
        }
      }
    }
    if (!data.pageInfo.hasNextPage) return matches;
    cursor = data.pageInfo.endCursor;
  }
}

async function applyCorrection(variantId, warehouseId, correctedQuantity) {
  const result = (await gql(STOCK_BULK_UPDATE, {
    variantId, warehouseId, quantity: correctedQuantity,
  })).stockBulkUpdate;
  for (const item of result.results) {
    if (item.errors.length) throw new Error(JSON.stringify(item.errors));
  }
  return result;
}

async function confirmAvailable(variantId, channel) {
  const data = (await gql(VARIANT_AVAILABILITY_QUERY, { id: variantId, channel })).productVariant;
  return data.quantityAvailable;
}

export async function run(knownPhysicalCounts = {}) {
  const stocks = await stockSnapshot(CHANNEL);
  const flagged = [];

  for (const stock of stocks) {
    const allocations = await allocationsFor(stock.variantId, stock.warehouseId);
    const known = knownPhysicalCounts[`${stock.variantId}::${stock.warehouseId}`] ?? null;
    const result = detectStockDrift(stock, allocations, known);
    if (!result.isDrift) continue;
    flagged.push({ ...stock, ...result, suspectedPhysicalCount: known });
    console.warn(
      `DRIFT sku=${stock.sku} variant=${stock.variantId} warehouse=${stock.warehouseId} quantity=${stock.quantity} allocated=${stock.quantityAllocated} reason=${result.reason} delta=${result.delta}`
    );
  }

  if (DRY_RUN) {
    console.log(`Done (dry run). ${flagged.length} drifted variant and warehouse pair(s) reported.`);
    return flagged;
  }

  for (const row of flagged) {
    if (row.suspectedPhysicalCount === null) {
      console.log(`Skipping ${row.sku} at ${row.warehouseId}, no confirmed physical count supplied.`);
      continue;
    }
    await applyCorrection(row.variantId, row.warehouseId, row.suspectedPhysicalCount);
    const available = await confirmAvailable(row.variantId, CHANNEL);
    console.log(`Corrected ${row.sku} at ${row.warehouseId} to ${row.suspectedPhysicalCount}. quantityAvailable now ${available}.`);
  }

  console.log(`Done. ${flagged.length} drifted variant and warehouse pair(s) processed.`);
  return flagged;
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
